design/one-pager-default-resource-class.md

```diff
@@ -55,21 +55,21 @@ While this provides a nice separation of concerns for the developer and the oper
 
 ## Proposed Workflow
 
-While it will remain possible to explicitly reference an underlying resource class, developers will now have the option to omit the class reference and rely on falling back to whatever operators have deemed an appropriate default. The default resource class will be distinguished via annotation:
+While it will remain possible to explicitly reference an underlying resource class, developers will now have the option to omit the class reference and rely on falling back to whatever operators have deemed an appropriate default. The default resource class will be distinguished via the `default` and `abstractResource` fields:
 
 ```yaml
 apiVersion: core.crossplane.io/v1alpha1
 kind: ResourceClass
 metadata:
   name: cloud-postgresql
   namespace: crossplane-system
-  annotations:
-    postgresql.storage.crossplane.io/default: "true"
 parameters:
   class: db.t2.small
   masterUsername: masteruser
   securityGroups: "sg-ab1cdefg,sg-05adsfkaj1ksdjak"
   size: "20"
+default: true
+abstractResource: postgresqlinstance.storage.crossplane.io
```

Comment:
At Upbound we've been thinking about how we might handle "non-portable" resources like AWS DynamoDB and Google Cloud Datastore. In this context "non-portable" means resources solve similar problems, but are different enough that there's no way we could bind them to a single kind of resource claim.

Let's assume these two resources were represented by `dynamodb.database.aws.crossplane.io` and `clouddatastore.database.gcp.crossplane.io` concrete resource kinds. These resources are both NoSQL databases, but have sufficiently different protocols and semantics that we probably could not claim them both using a single `nosql.storage.crossplane.io` resource claim. Instead we'd probably establish new claims that could only be satisfied by a single resource kind (for example `dynamodb.database.crossplane.io` would map only to `dynamodb.database.aws.crossplane.io`). This sacrifices multicloud portability, but allows us to maintain Crossplane's separation of concerns in that SREs could configure a few resource classes for DynamoDB, including a default resource class, freeing developers from worrying about the specific configuration of their DynamoDB databases.

If we take this idea a little further, it stands that we might have two different claim kinds that could be satisfied by the `cloudmemorystoreinstance.cache.gcp.crossplane.io` concrete resource; `rediscluster.cache.crossplane.io`, and `cloudmemorystoreinstance.cache.crossplane.io`. In this scenario I could imagine we might have one resource class that could satisfy two different resource claim kinds. I wonder if we could represent this with a single slice like:

```yaml
apiVersion: core.crossplane.io/v1alpha1
kind: ResourceClass
metadata:
  name: default-rediscluster
  namespace: crossplane-system
// ... Other resource class fields
defaultForClaimKinds:
- rediscluster.cache.crossplane.io
- cloudmemorystoreinstance.cache.crossplane.io
```

Reply:
I was looking for annotation vs field justification and dug this up:
https://github.com/kubernetes/kubernetes/pull/31617#issuecomment-248281559

> In the future, we hope to create a real StorageClass attribute with validation and admission controller that lets only one class to be marked as default.

Reply:
I definitely like this functionality and agree that it plays well with non-portable resources. Thanks to both of you for feedback! I will update in next commit.